libbeat/processors/actions/truncate_fields.go

```diff
@@ -0,0 +1,180 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package actions
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+type truncateFieldsConfig struct {
+	Fields        []string `config:"fields"`
+	MaxBytes      int      `config:"max_bytes" validate:"min=0"`
+	MaxCount      int      `config:"max_character_count" validate:"min=0"`
+	IgnoreMissing bool     `config:"ignore_missing"`
+	FailOnError   bool     `config:"fail_on_error"`
+}
+
+type truncateFields struct {
+	config   truncateFieldsConfig
+	truncate truncater
+}
+
+type truncater func(*truncateFields, []byte) ([]byte, error)
+
+func init() {
+	processors.RegisterPlugin("truncate_fields",
+		configChecked(newTruncateFields,
+			requireFields("fields"),
+			mutuallyExclusiveRequiredFields("max_bytes", "max_count"),
+		),
+	)
+}
+
+func newTruncateFields(c *common.Config) (processors.Processor, error) {
+	var config truncateFieldsConfig
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the truncate_fields configuration: %s", err)
+	}
+
+	var truncateFunc truncater
+	if config.MaxBytes > 0 {
+		truncateFunc = (*truncateFields).truncateBytes
+	} else {
+		truncateFunc = (*truncateFields).truncateCharacters
+	}
+
+	return &truncateFields{
+		config:   config,
+		truncate: truncateFunc,
+	}, nil
+}
+
+func (f *truncateFields) Run(event *beat.Event) (*beat.Event, error) {
+	var backup common.MapStr
+	if f.config.FailOnError {
+		backup = event.Fields.Clone()
+	}
+
+	for _, field := range f.config.Fields {
+		event, err := f.truncateSingleField(field, event)
+		if err != nil && f.config.FailOnError {
+			logp.Debug("truncate_fields", "Failed to truncate fields: %s", err)
+			event.Fields = backup
+			return event, err
+		}
+	}
+
+	return event, nil
+}
+
+func (f *truncateFields) truncateSingleField(field string, event *beat.Event) (*beat.Event, error) {
+	v, err := event.GetValue(field)
+	if err != nil {
+		if f.config.IgnoreMissing && errors.Cause(err) == common.ErrKeyNotFound {
+			return event, nil
+		}
+		return event, fmt.Errorf("could not fetch value for key: %s, Error: %+v", field, err)
+	}
+
+	switch value := v.(type) {
+	case string:
+		return f.addTruncatedString(field, value, event)
+	case []byte:
+		return f.addTruncatedByte(field, value, event)
+	default:
+		return event, fmt.Errorf("value cannot be truncated: %+v", value)
+	}
+
+}
+
+func (f *truncateFields) addTruncatedString(field, value string, event *beat.Event) (*beat.Event, error) {
+	truncated, err := f.truncate(f, []byte(value))
```

Comment:
Converting from `string` -> `[]byte` -> `string` going to do two separate copies of the string data, but since it's only truncating I think this could be avoided with some refactoring. It would be good if someone could double-check my assumption about the two copies.

Reply:
It is indeed copying twice. But after thinking a bit more about the conversion and copying data I am not sure if it makes sense to eliminate them. First I need to copy and convert the string, as it is inmutable type, in order to modify it. Then I need to create a new truncated copy, so the previous long array can be freed. Then as I need to return a string, a new copy is needed to get the inmutable type again.
Without converting the string to []byte, I can only do slicing which does not touch the underlying data, thus the long line stays in memory as long there is a reference to it. I think it would take away the advantage of decreasing the size of the data in memory.
WDYT?

Reply:
I also think that we would mostly truncate `[]byte` fields (e.g. message) and rarely on string. Thus, this code path would be less frequently run.

Reply:
Also, I can reduce the number of copies by not passing around the arrays. Will do that.